Issue#21 advanced search

thineshth opened the pull request:
Fixes #21 - Advanced Search functionality

thineshth committed:
issue-#21-AdvancedSearch:

- Implemented rudimentary advanced search logic with basic UI.
-Needs testing on slower CPUs to make sure they way I'm passing info between activities isn't too slow and memory intensive.
- Added toString override function for types
- Added a getAllTypes in PokemonDBAccess

thineshth committed:
issue-#21-AdvancedSearch:

- Pushing missing manifest file

thineshth committed:
issue-#21-AdvancedSearch:

- Two type search implemented.
- Put in the safeguard so that all minimal pokemone are loaded before let into filter.

thineshth committed:
issue-#21-AdvancedSearch:

- Advanced Search done for EggGroups/Types/Ability/Generation Available
- Added appropriate functions to PokemonFactory to access

thineshth committed:
issue-#21-AdvancedSearch:

- Fixed a bug where filters weren't reset after pressing back
- Fixed a bug where eggGroupID started at 1, when it starts at 0 in DB

thineshth committed:
issue-#21-AdvancedSearch:

- Currently won't build due to onFocusChange function

thineshth committed:
issue-#21-AdvancedSearch:

- Finished functions for filtering by base stats in AdvancedSearch and the corresponding in PokemonFactory
- Basic UI without color is implemented
- Missing is filtering by Total base stats, currently investigating the SQLlite queries required

thineshth committed:
issue-#21-AdvancedSearch:

- Merged with Master. So overlapping functions in PokemonFactory were resolved

foxtrotzulu94 commented on app/src/main/res/values/strings.xml:
```diff
@@ -30,16 +30,6 @@
     <!-- DetailedPokemonActivity Strings End-->
 
-
-    <!-- PokemonResources Strings Start-->
```
The branch is currently not building due to these missing strings.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/pokemon/EggGroup.java:
```diff
@@ -51,4 +51,14 @@ public EggGroup(String name) {
+     */
+    @Override
+    public String toString(){
+        return String.format("%s",name);
```
You could just return name directly.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/pokemon/Ability.java:
```diff
@@ -42,4 +42,14 @@ public String getName() {
+     */
+    @Override
+    public String toString(){
+        return String.format("%s",name);
```
Return name directly.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -390,7 +406,7 @@ private void minimalPokemonNameInitializer(Cursor cursor){
 
 
     public void getAllDetailedPokemon(){
-        //NOTE: This method is here for testing purposes. Avoid using in production builds.
```
This comment should be preserved since it is unlikely we'll build all Pokemon objects in one shot.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1154,79 @@ public void loadAllTypes() {
+        // get the real abilities from the database:
+        for (int i = 1; i < getMaxAbilityID() + 1; i++) {
+            String[] selectionArg = {String.valueOf(i)};
+            Cursor cursor = database.query(ABILITIES_TABLE, null, ABILITIY_ID + "=?", selectionArg, null, null, null);
```
Consider querying the entire table first and then using a "do{..}while()" loop to read from the cursor all rows retrieved

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1154,79 @@ public void loadAllTypes() {
+            Cursor cursor = database.query(ABILITIES_TABLE, null, ABILITIY_ID + "=?", selectionArg, null, null, null);
+            cursor.moveToFirst();
+
+            abilities[i] = new Ability(cursor.getString(cursor.getColumnIndex(NAME)), cursor.getString(cursor.getColumnIndex(DESCRIPTION)));
```
Similarly, after doing a single query, the columns won't change, so you can store the result of cursor.getColumnIndex in an integer. This will make your loop considerably faster.

minicole replied:
Consider adding the abilityID as a private variable to the Ability class (like with Type). This will give easier access to the index in the array. (Same below with EggGroup)

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1154,79 @@ public void loadAllTypes() {
+        for (int i = 1; i < getMaxEggGroupID() + 1; i++) {
+            // eggGroupId's start at 0
+            String[] selectionArg = {String.valueOf(i - 1)};
+            Cursor cursor = database.query(EGG_GROUPS_TABLE, null, EGG_GROUP_ID + "=?", selectionArg, null, null, null);
```
Same comment as with abilities. Make a single query, read each row and only ask for column index once.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1154,79 @@ public void loadAllTypes() {
+     * Loads all the National Id's into memory. The position+1 would be the unique ID
+     * Should be done at program start (during splash screen)
+     */
+    public void loadAllUniqueToNationalID() {
```
I wrote similar code when loading all MinimalPokemon. If you use all the indexes of this array, it would be good to refactor those methods so that the same work is done only once. Also, this is by far faster to setup and lookup than the HashMap I was used.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1138,6 +1243,224 @@ private int getMoveID(Move move) {
+
+    public ArrayList<Integer> getAllUniqueIDsFromAbility(int abilityID){
+
+        int actualAbilityID = abilityID;
```
actualAbilityID is not written to again. Consider using abilityID directly.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1138,6 +1243,224 @@ private int getMoveID(Move move) {
+//        else{
+//            // Filter by the total sum of base stats
+//            // using a raw query to get sum, not sure if even possible with query
+//            Cursor cursor = database.rawQuery("select (HP + ATTACK + DEFENCE + SPATTACK + SPDEFENCE + SPEED) as sum from pokemon_unique_info where sum >="+String.valueOf(lowerLimit)+" AND sum <="+String.valueOf(upperLimit)+
```
We should probably precalculate and place the Stat Sum in the DB. This should avoid the runtime calculation.

minicole replied:
You can put a comment about this in the Issue #43 Scraper wraper task.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,512 @@
+ */
+
+
+public class AdvancedSearch extends ActionBarActivity {
```
Refactor this to "AdvancedSearchActivity extends AppCompatActivity". ActionBarActivity is now deprecated.

thineshth committed:
issue-#21-AdvancedSearch:

- Optimized AdvancedSearch by caching max_unique_id and max_national_id in PokemonFactory

- Fixed missing string files in strings.xml

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+ * Generation Available
+ * Egg Groups 1
+ * Egg Groups 2
+ * TODO: All 6 Base stats
```
is this done? maybe leave TODO: sum

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+    PokemonFactory pokemonFactory;
+
+    // Pokemon information to filter
+    MinimalPokemon[] allMinimalPokemon;
```
Never used.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+    MinimalPokemon[] allMinimalPokemon;
+
+    //Final Filtered List
+    MinimalPokemon [] filteredMinimalPokemon;
```
Never used.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+    Type[] allTypes;
+    Ability[] allAbilities;
+    EggGroup[] allEggGroups;
+    String allGenerations [];
```
Shouldn't it be:
String[] allGenerations;

thineshth replied:
Hmm, seems to be working. But I'll change it. I'm not sure why I had it like that.

minicole replied:
According to [stackoverflow](http://stackoverflow.com/questions/1200621/declare-array-in-java) both work, but [] after String is preferred. So your choice.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_advanced_search);
+
+        pokemonFactory = PokemonFactory.getPokemonFactory(this.getApplicationContext());
```
The PokemonFactory singleton is already instantiated at the start of the application, so you can pass null here.

thineshth replied:
By not passing null, does it do extra work?

minicole replied:
A bit. It's calling this.getApplicationContext(). I don't know how heavy that method is, or how big the context is. Other than that, it's not going to hurt or confuse the PokemonFactory because it will just retrieve the singleton instance that already exists.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+
+        pokemonFactory = PokemonFactory.getPokemonFactory(this.getApplicationContext());
+
+        allTypes = pokemonFactory.getListOfTypes();
```
Since these four methods return the same type of thing, they should be named similarly.

thineshth replied:
Renaming getListOfTypes() -> getAllTypes().

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+        Log.d("QPDex", String.format("Obtaining Spinner Data took %s ns",duration));
+
+        // No need for MinimalPokemon
+        //allMinimalPokemon = pokemonFactory.getAllMinimalPokemon();
```
remove unnecessary commented code (unless you might need it in the future).

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+
+    private void fillInSpinnerValues(){
+
+        //TODO: Filling out these Spinners with the actual objects seems to be causing a bottleneck,
```
Is this a bottleneck that should be fixed by Milestone 1?

thineshth replied:
Removing this comment, the actual bottleneck was at the DB access level. Filling in the spinners is actually quite fast.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+            // parent.getItemAtPosition(pos)
+
+            // Determine other type, and set current type
+            Type otherType = new Type("Temp", -1);
```
I'm not clear on how otherType is used. It is a local variable to this method that is never read.
Is it meant to make it so that you can't select the same type twice?

thineshth replied:
It was used for that. But it's no longer needed since all the filtering occurs right before intent is switched. Removing it.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+        collectStats();
+
+        //Comment this out, takes way tooo long 4+sec
+        currentFilterPokemonArrayList.clear();
```
Is this meant to be commented out?

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+
+        if(selectedType1.getTypeID() != 0){
+
+            tempFilterPokemonArrayList = pokemonFactory.getAllUniqueIDsFromType(Arrays.asList(allTypes).indexOf(selectedType1));
```
Consider using selectedType1.getTypeID() and selectedType2.getTypeID()

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+        //Start with the ability chosen, if any. Use the ability ID to get all NationalID's associated
+        if(selectedAbility.getName() != "None"){
+
+            currentFilterPokemonArrayList = pokemonFactory.getAllUniqueIDsFromAbility(Arrays.asList(allAbilities).indexOf(selectedAbility));
```
if you put abilityID as a variable in Ability class (as commented somewhere below), you don't have to turn the array into a list to find the index.

thineshth replied:
I agree with this refactoring. It should definitely be done, it makes life a lot easier.

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+        filteredNationalIds = pokemonFactory.convertUniqueToNational(currentFilterPokemonArrayList);
+
+        // Filter national IDs based on Egg Groups
+        if(selectedEggGroup1.getName() != "None"){
```
Should be: 
if (!selectedEggGroup1.getName().equals("None")) {
With strings in Java, use blabla.equals("") instead of == or !blabla.equals("") instead of !=

Same for selectedEggGroup2 below

minicole commented on app/src/main/java/me/quadphase/qpdex/AdvancedSearch.java:
```diff
@@ -0,0 +1,508 @@
+        // Filter national IDs based on Egg Groups
+        if(selectedEggGroup1.getName() != "None"){
+
+            tempFilterPokemonArrayList = pokemonFactory.getAllNationalIdsFromEggGroup(Arrays.asList(allEggGroups).indexOf(selectedEggGroup1));
```
same as above with eggGroupID (like ability)

minicole commented on app/src/main/java/me/quadphase/qpdex/PokedexActivity.java:
```diff
@@ -171,7 +176,34 @@ public void onClick(View view) {
+
+            ArrayList<Integer> filteredPokemonNationalid = extraInfo.getIntegerArrayList("FILTERED_POKEMON_NATIONALID");
+
+            MinimalPokemon [] filteredlist = new MinimalPokemon[filteredPokemonNationalid.size()];
```
[MINOR]: Whitespace before brackets.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -108,6 +108,12 @@
+    private EggGroup[] eggGroups;
+    private Ability[] abilities;
+    private int[] allNationalIDMappedUnique;
+    private  String[] generations;
```
[MINOR]: whitespace between private and String

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -390,13 +412,13 @@ private void minimalPokemonNameInitializer(Cursor cursor){
 
     public void getAllDetailedPokemon(){
-        //NOTE: This method is here for testing purposes. Avoid using in production builds.
+        //TODO: Perform aggressive optimizations when possible
```
aggressive optimizations have been performed. (remove TODO)

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -892,7 +914,7 @@ public void setCaught(int nationalID, boolean value){
      */
     public int getCurrentGeneration() {
```
this method does what getMaxGeneration() does, and should therefore be safely deleted.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1196,79 @@ public void loadAllTypes() {
+            Cursor cursor = database.query(EGG_GROUPS_TABLE, null, EGG_GROUP_ID + "=?", selectionArg, null, null, null);
+            cursor.moveToFirst();
+
+            eggGroups[i] = new EggGroup(cursor.getString(cursor.getColumnIndex(NAME)));
```
Consider adding the eggGroupID as a private variable to the EggGroup class (like with Type). This will give easier access to the index in the array.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1138,6 +1285,215 @@ private int getMoveID(Move move) {
+
+        int actualAbilityID = abilityID;
+        ArrayList<Integer> filteredUniqueids = new ArrayList<>();
+        ArrayList<Integer> filteredNationalids = new ArrayList<>();
```
never used.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1138,6 +1285,215 @@ private int getMoveID(Move move) {
+    public ArrayList<Integer> getAllUniqueIDsFromType(int typeID){
+
+        ArrayList<Integer> filteredUniqueids = new ArrayList<>();
+        ArrayList<Integer> filteredNationalids = new ArrayList<>();
```
never used.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1138,6 +1285,215 @@ private int getMoveID(Move move) {
+     * @return array of integers of the unique id's that correspond
+     */
+    public ArrayList<Integer> getAllUniqueIDsFromStat(int lowerLimit, int upperLimit, String stat){
+        //TODO: Add validation step for stat string
```
Reminder TODO!

Consider instead of a string, make it an enum with the possible values.

thineshth replied:
I'm having trouble with enums, for now I added validation with conditions and changed the todo

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1196,79 @@ public void loadAllTypes() {
+    public void loadAllGenerations() {
+
+        // fail safe 0th index of generations:
+        generations[0] = "-";
```
Consider "gen" because it is unclear what the dropdown is filtering for.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1196,79 @@ public void loadAllTypes() {
+     */
+    public void loadAllEggGroups() {
+        // fail safe 0th index of types:
+        eggGroups[0] = new EggGroup("None");
```
Consider "Egg Group" since it is unclear what each of the filters are.

minicole commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1122,6 +1196,79 @@ public void loadAllTypes() {
+     */
+    public void loadAllAbilities() {
+        // fail safe 0th index of types:
+        abilities[0] = new Ability("None","None");
```
Consider naming this "Ability" since it is unclear what each of the filters are.

minicole commented:
1. Possible bug: Search of "Dark" "Dark" gives 130. Gyarados (Flying, Water) and 493. Arceus (Normal) and 720. Hoopa (Ghost, Psychic) along with the dark types. (this is where a test suite would be super nice)

2. Consider making the first element of the type lists Type instead of "None/Bird". (Same for other filters, which I noted with inline comments).

3. Consider putting a background more similar to the other pages to make the app seem more "unified". (Up for debate)

4. Consider that if the list of filtered pokemon is empty, do not go to PokedexActivity and just display a modal/popup/message indicating that "No Pokemon match this selection" or something.

thineshth commented:
For 1. It's showing those pokemon under Dark because either their Mega evolution or their alternative form has dark type.

2. I agree, and makes it easier than putting a title above the box

3. Considering putting the grid background with the a different color? Any ideas which?

4. Sounds like a good idea

thineshth committed:
issue-#21-AdvancedSearch:

- Added Total Base stats to the pokemon_unique_info table and filled it out. Made changes for scrapers.

- Added greenish tile for Advanced Search and fixed up layout

- Made code review changes to finish off pull request

thineshth committed:
Merge branch 'master' into Issue#21-AdvancedSearch

thineshth committed:
issue-#21-AdvancedSearch:

- Added custom xml background box to advanced search activity layout
- Added total base stats as a string value in strings.xml
- Adjusted relative layout to make pretty

thineshth committed:
issue-#21-AdvancedSearch:

- Refactored use of hashmap to map unique to national ids to an array: allNationalIDMappedUnique
